### PR TITLE
Nearly equal with relative and absolute tolerance

### PR DESCRIPTION
**Pull request by dvd101x**
Hi, 

This initial commit is intended for exploration of #2838. It overwrites the current `nearlyEqual` completely to be as similar to python's [Math.isclose](https://docs.python.org/3/library/math.html#math.isclose) as possible (according to the description in the documentation)

It currently fails 6 tests:
- 2 tests in `rotationMatrix`
- 1 test in `nearlyEqual` (for bigNumber)
- 3 tests in `number / nearlyEqual`

No effort was made to either adapt the tests or fix the function. Everything is left as is to review the failing tests and aid in the decision making.

Please note that the functions are changed abruptly only for exploration and can be returned to their original form upon review.

Also no calls for `nearlyEqual` were modified, thus only relative tolerance is being used. If the default values are changed to `relTol = Number.EPSILON, absTol = 1e-12` then a few more tests fail.

### CODE REVIEW

**Comment by josdejong** [comment]
Thanks, this is a great start! 

Some thoughts:

1. Only 8 failing unit tests is not that much :). I don't see failing tests related to `rotationMatrix` though, I see 4 tests for `eigs` failing?
2. I expect more tests to fail as soon as we update all places where `nearlyEqual(...)` is invoked to pass both `relTol` and `absTol`.
3. I think we may need to add some additional unit tests to test `relTol` and `absTol` individually.
4. We should decide upon good default values `relTol` and `absTol`, and how we map values for both when a user configures `epsilon` (for backward compatibility). Would it make sense to default to `relTol = epsilon ?? 1e-12` and `absTol = epsilon ?? 1e-12`?
5. In general I like to use full words rather than abbreviations, like `relativeTolerance` and `absoluteTolerance`. However in this case the words are quite long. Still thinking about that.
6. I see the tests where run on node.js 16, I don't understand why, it should only run node 18 and 20 ([see config](https://github.com/josdejong/mathjs/blob/develop/.github/workflows/build.yaml#L16))? Is something in this branch not up to date?

**Comment by dvd101x** [comment]
1. That's odd, I don't see `eigs` failing, maybe this is not up to date.
2. I agree
3. I agree
4. I think the current behavior is similar to having an `absTol` about 1000 times smaller than `relTol`. I see on the julia docs for [isaprox](https://docs.julialang.org/en/v1/base/math/#Base.isapprox) some ways to calculate those default values.
5. I agree these names don't match mathjs very well just took the python names and got a warning related to the camel case. I'm sure there are more appropriate names. Regarding abbreviations it could be argued that `config` is an abbreviation.
6. At some point I saw that line trying to update in my PR but I returned it to what it was. Probably is a mistake I did.

**Comment by dvd101x** [comment]
A quick comparison for names and values.

| language | function | relative tolerance | absolute tolerance |
|---|---|----|---|
|python | [isclose](https://docs.python.org/3/library/math.html#math.isclose)| rel_tol = 1e-09 | abs_tol = 0.0 |
|numpy| [isclose](https://numpy.org/doc/stable/reference/generated/numpy.isclose.html#numpy-isclose) | rtol = 1e-05 | atol = 1e-08
| julia | [isapprox](https://docs.julialang.org/en/v1/base/math/#Base.isapprox) | rtol = atol>0 ? 0 : √eps  | atol = 0 |

**Comment by josdejong** [comment]
(1) You can see the test results by clicking them in the list with checks in this PR, for example https://github.com/josdejong/mathjs/actions/runs/7853061106/job/21432241083?pr=3152. I guess indeed that you somehow have an old version or so (that may also explain why the PR is still testing on Node 16).

(5) ha ha, that's true. I'm ok with either `absTol / relTol` or `absoluteTolerance / relativeTolerance`, I don't have a very strong opinion in this regard (in general, I like the non abbreviated name, but in this case it is _quite_ long).

**Comment by josdejong** [comment]
I think the next step is to work out this PR in detail:  update all places where `nearlyEqual(...)`  is invoked, and then fix all failing unit tests.

**Comment by dvd101x** [comment]
Ok, thanks! I will do that and report back.

**Comment by dvd101x** [comment]
Hi, just reporting back.

Progress has been slower than what I originally expected but I'm still on it.

**Comment by josdejong** [comment]
Thanks for the update David, good to know.

**Comment by dvd101x** [comment]
Hi Jos

I have some good news!

I found a good fit with a relative tolerance the size of the current `config.epsilon` 1e-12 and an absolute tolerance of 1e-15. I did some empirical tests and works similarly in most cases.

In the latest submit I made all calls for `nearlyEqual` to use `relTol  = config.epsilon` and `absTol = config.epsilon * 1e-3` and this **eliminates all the failing tests** except for the `nearlyEqual` tests.

I would like your advise with the following:
1. Shall I remove `config.epsilon` in favor of `config.relTol` and `config.absTol`? (naming to be defined)
2. Shall I change the failing `nearlyEqual` tests to accommodate for the proposed `nearlyEqual` or make a compromise between the proposed `nearlyEqual` and the current tests?
3. Shall I make a mathjs function `nearlyEqual` available so that the user can choose `absTol` and `relTol` without changing config or shall it be a different topic outside this PR?

# Appendix

Here is the mathjs code I used to test for the tolerances: (this works on the current version **12.4.1**
```jl
config({epsilon:1e-12});

# Define function nearly equal
nearlyEqual(a, b, relTol, absTol) = abs(a-b) <= max(relTol * max(abs(a), abs(b)), absTol);

# Define rtol and atol 
rtol = config().epsilon;
atol = rtol * 1e-3;

# Tested various values of x
X = [0, 0.1, 0.5, 1, 10, 1e3, 1e6, 1e12];
x = X[1];

# Ys
powers = -2:-1:-16;
Y = x + map(powers, _(d) = pow(10, d));

# Proposed
nearlyEqualResults = map(Y, _(y) = nearlyEqual(x, y, rtol, atol));

# Current
equalResults = x == Y;

# These are the powers where the current nearly equals differ from the proposed one
differences = nearlyEqualResults != equalResults;

result = {
  powersFailing:powers[differences],
  x:x,
  current:equalResults[differences],
  proposed:nearlyEqualResults[differences],
  rtol:rtol,
  atol:atol
};

print("Testing with:\n
rtol: $rtol
atol: $atol
x: $x\n
yields:
powers faling: $powersFailing
proposal: $proposed
current: $current
", result)
```

I read the following document to realize that `nearlyEqual` might need to be one of the mathjs functions.
- [MDN: Number.EPSILON](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/EPSILON)

**Comment by josdejong** [comment]
Yay, that sounds promising!

1. Yes, though we do need backward compatibility in the function `config()`. When people are still using `.epsilon`, we can translate it internally into the new options `relTol` and `abstol` and output a console warning about the deprecation.
2. I think we have to adjust the failing `nearlyEqual` tests. This will be a breaking change, and I suppose the failing tests are due to the behavior now being different, so the tests should be adapted accordingly. (please correct me if I'm wrong)
3. I'm not sure if we should expose the `nearlyEqual` function, it may make sense 🤔. I prefer to discuss that in a separate topic to keep this PR focused, OK?

Looking at the code example, you could definitely use #2953 🤣

**Comment by dvd101x** [comment]
Thanks!

1. **OK:** sounds good, I was worried it was too much change.
2. **OK:** most are due to the new formula. There is one that might be debatable that says `nearlyEqual: should do exact comparison when epsilon is null or undefined`. In my opinion it should `resort to default values of relTol and absTol when they are null or undefined`
3. **OK:** Will make a separate discussion.

Absolutely! I was trying to change my point of view about the shorthand and I did.
> **PD:** Also could have used #2701 to cycle through the values of X

**Comment by josdejong** [comment]
👍 

I see you're pushing new commits. Please let me know when the PR is ready for review (or if you hit new challenges).

**Comment by dvd101x** [comment]
Hi Jos,

Thanks for the follow up.

I'm still tweaking some stuff and will let you know when it's ready for review.

**Comment by dvd101x** [comment]
Hi Jos, this is ready for review

**Comment by josdejong** [comment]
Thanks, that is good news! I'll look into it asap.

File: src/core/function/config.js [hunk 1]
```diff
@@ -49,23 +52,33 @@ export function configFactory (config, emit) {
    */
   function _config (options) {
     if (options) {
-      const prev = mapObject(config, clone)
+      if (options.epsilon !== undefined) {
+        // backwards compatibility
+        console.warn('Warning: The configuration option "epsilon" is deprecated. Use "relTol" and "absTol" instead.')
+        const optionsFix = mapObject(options, clone)
```
Comment (josdejong) [inline src/core/function/config.js:58]: I think this can be simplified to `const optionsFix = clone(options)`

File: src/core/function/config.js [hunk 2]
```diff
@@ -49,23 +52,33 @@ export function configFactory (config, emit) {
...
+        optionsFix.relTol = options.epsilon
+        optionsFix.absTol = options.epsilon * 1e-3
+        delete optionsFix.epsilon
+        return _config(optionsFix)
+      } else {
```
Comment (josdejong) [inline src/core/function/config.js:63]: Can you remove the `else` case, so we keep the "old" as is and only add a small `if() {...}` section for backward compatibility that we can remove some day without having to alter the rest of the code?

File: types/index.d.ts [hunk 3]
```diff
@@ -4296,7 +4298,8 @@ export interface Help {
 }
 
 export interface ConfigOptions {
-  epsilon?: number
+  relTol?: number
+  absTol?: number
```
Comment (josdejong) [inline types/index.d.ts:4302]: I think we need to keep `epsilon` here and mark it as deprecated, else existing TypeScript code that uses `epsilon` will break (and the backward compatibility doesn't work).

File: test/unit-tests/utils/number.test.js [hunk 4]
```diff
@@ -505,83 +505,97 @@ describe('number', function () {
...
 
-      assert.strictEqual(nearlyEqual(0.1 + 0.2, 0.3), false)
-      assert.strictEqual(nearlyEqual(0.1 + 0.2, 0.3, null), false)
+      assert.strictEqual(nearlyEqual(0.1 + 0.2, 0.3), true)
+      assert.strictEqual(nearlyEqual(0.1 + 0.2, 0.3, undefined), true)
+      assert.strictEqual(nearlyEqual(0.1 + 0.2, 0.3, undefined, undefined), true)
+
+      assert.strictEqual(nearlyEqual(1.2 + 1e-7, 1.2), false)
+      assert.strictEqual(nearlyEqual(1.2 + 1e-7, 1.2, undefined), false)
+      assert.strictEqual(nearlyEqual(1.2 + 1e-7, 1.2, undefined, undefined), false)
```
Comment (josdejong) [inline test/unit-tests/utils/number.test.js:598]: Can you add a new unit test that validates the backward compatibility, using the deprecated `epsilon`?

**Review by josdejong — CHANGES_REQUESTED** [review]
This looks really solid David, thanks! 

Hardly any existing unit test breaks with the new behavior, I'm kind of (happily) surprised by that. I had expected some changes with edge cases of the rounding functions. Or am I over looking something?

I made a couple of inline comments, can you have a look at those?

**Comment by dvd101x** [comment]
Hi Jos, thanks for the thorough review!

I'm also glad there where not so many errors. The reason might be that the values for the new tolerances were picked to be close to the existing behavior. Maybe at some point in the future the values might change and there will be more discrepancies but at least there is the option to adjust as needed case by case.

I will work on the inline comments and report back probably during the weekend.

File: src/core/function/config.js (hunk 1 above)
Comment (dvd101x) [reply src/core/function/config.js:58]: O that's right, it works also for all instances of mapObject, so mapObject could be removed from importrs.

File: src/core/function/config.js (hunk 2 above)
Comment (dvd101x) [reply src/core/function/config.js:63]: Thanks, got it. I thougt it could look more common with the else. But I guess the diff will look cleaner in the future.

File: test/unit-tests/utils/number.test.js (hunk 4 above)
Comment (dvd101x) [reply test/unit-tests/utils/number.test.js:598]: Yes of course,

I did include them on `config.test.js` but found an issue with console.warn that I don't know how to test and also breaks the pattern of the tests that now looks like this.

```
.....................................Warning: The configuration option "epsilon" is deprecated. Use "relTol" and "absTol" instead.
.....................,,,,,.......................
  ......................................................................................
```

This is my error but I don't know how to fix.

File: types/index.d.ts (hunk 3 above)
Comment (dvd101x) [reply types/index.d.ts:4302]: Ok, just did.

How do I mark it as deprecated?

**Comment by dvd101x** [comment]
Hi Jos, it's ready for review again. I included some inline comments with a few questions.

File: types/index.d.ts (hunk 3 above)
Comment (josdejong) [reply types/index.d.ts:4302]: Like this:

```ts
/**
 * @deprecated Use `relTol` and `absTol` instead
 */
epsilon?: number
```

File: test/unit-tests/utils/number.test.js (hunk 4 above)
Comment (josdejong) [reply test/unit-tests/utils/number.test.js:598]: Ah, yes. We can do two things:

1. Integrate and use some spy library for mocha, like [Sinon](https://sinonjs.org/releases/latest/spies/)
2. Temporarily replacing `console.warn` with our own custom function, and after the test restore the original `console.warn`

Option 1 would be best, but only if it is simple to setup, otherwise option 2 is fine with me. Option 2 is a bit ugly: when the test throws an exception the original console.warn is not restored, therefore it's not my first choice.

**Review by josdejong — COMMENTED** [review]
Thanks for the updates! I've replied to the two left over comments, can you have a look at them again?

File: test/unit-tests/utils/number.test.js (hunk 4 above)
Comment (dvd101x) [reply test/unit-tests/utils/number.test.js:598]: Thanks!

I will research sinon and report back.

Comment (dvd101x) [reply test/unit-tests/utils/number.test.js:598]: Hi, it seems to be working now, please review.

File: types/index.d.ts (hunk 3 above)
Comment (dvd101x) [reply types/index.d.ts:4302]: Thanks!
It's done now.

**Comment by dvd101x** [comment]
Hi Jos, the two left comments were addressed.

Please review.

File: test/unit-tests/utils/number.test.js (hunk 4 above)
Comment (josdejong) [reply test/unit-tests/utils/number.test.js:598]: Nice job, thanks!

**Review by josdejong — APPROVED** [review]

**Comment by dvd101x** [comment]
Thanks Jos!

Looking forward to `v13`!

**Comment by josdejong** [comment]
Published now in `v13.0.0` 🎉
